### PR TITLE
Adds in QuikDeploy:tm: barricades

### PR DESCRIPTION
**Pull request by ma44**
## About The Pull Request
Admin spawn only item that allows for constructing a plasteel or metal barricade in two seconds, the item is smaller than a stack of metal/plasteel but less efficient for backpack space taken => amount of barricades made

## Why It's Good For The Game
Allows for extremely quick FOB construction for some events if admins wanna use it

## Changelog
:cl: ma44
add: Adds in QuikDeploy barricades, items that allow for quickly constructing barricades via the power of unfolding, pray to your local god and hope you get some.
/:cl:

### CODE REVIEW

File: code/game/objects/items/quickdeploy_cade.dm [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+/obj/item/quikdeploy_cade
+	name = "QuikDeploy Barricade"
+	desc = "This is a QuikDeploy barricade, allows for extremely fast laying of barricades. Usually found in specific vendors and cannot be picked back up."
+	icon = 'icons/obj/items/quikdeploy_cade.dmi'
+	icon_state = "metal"
+	var/obj/thing_to_deploy = /obj/structure/barricade/metal
+	w_class = WEIGHT_CLASS_SMALL //While this is small, normal 50 stacks of metal is NORMAL so this is a bit on the bad space to cade ratio
+
+/obj/item/quikdeploy_cade/examine(mob/user)
+	..()
```
Comment (Rohesie) [inline code/game/objects/items/quickdeploy_cade.dm]:
```suggestion
	. = ..()
```

File: code/game/objects/items/quickdeploy_cade.dm [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	to_chat(user, "This QuikDeploy cade seems to deploy a [thing_to_deploy.name].")
+
+/obj/item/quikdeploy_cade/attack_self(mob/user)
+	to_chat(user, "You start to deploy the cade on your tile in front of you...")
+	if(!do_after(usr, 2 SECONDS, TRUE, src, BUSY_ICON_BUILD))
+		to_chat(user, "You decide against putting the barricade here.")
+		return
+	var/check_build = TRUE
+	for(var/obj/thing in get_turf(src))
+		if(!istype(thing, thing_to_deploy))
```
Comment (Rohesie) [inline code/game/objects/items/quickdeploy_cade.dm]: It should check for dense turfs, dense non-on_border objects, and dense on_border objects with the same dir. Else you can stack different types of barricades and windows.

File: code/game/objects/items/quickdeploy_cade.dm [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+		if(thing.dir != user.dir)
+			continue
+		check_build = FALSE
+		break
+	if(!check_build)
+		to_chat(user, "<span class='warning'>You can't deploy the barricades on top of another!</span>")
+		return
+	var/obj/O = new thing_to_deploy(get_turf(user))
+	O.setDir(user.dir)
+	playsound(loc, 'sound/items/ratchet.ogg', 25, 1)
```
Comment (Rohesie) [inline code/game/objects/items/quickdeploy_cade.dm]:
```suggestion
	playsound(loc, 'sound/items/ratchet.ogg', 25, TRUE)
```

File: code/game/objects/items/quickdeploy_cade.dm [hunk 4]
```diff
@@ -0,0 +1,37 @@
+//An item that allows a very quick deployment of a barricade
+/obj/item/quikdeploy_cade
```
Comment (psykzz) [inline code/game/objects/items/quickdeploy_cade.dm]:
```suggestion
/obj/item/quikdeploy/cade
```

You could make this a whole set of `QuikDeploy:tm:` items.

Optional, and up to you though

Comment (Rohesie) [reply code/game/objects/items/quickdeploy_cade.dm]: Good suggestion.

File: code/game/objects/items/quickdeploy_cade.dm (hunk 2 above)
Comment (ma44) [reply code/game/objects/items/quickdeploy_cade.dm]: Addressed that

Comment (Rohesie) [reply code/game/objects/items/quickdeploy_cade.dm]: Here's an example on how to do it: https://github.com/tgstation/TerraGov-Marine-Corps/pull/3566/files#diff-5b80535f483d4c008a0b2c4532a29f15R51

File: code/game/objects/items/quickdeploy_cade.dm [hunk 5]
```diff
@@ -0,0 +1,50 @@
+//An item thats meant to be a template for quickly deploying stuff like barricades
+/obj/item/quikdeploy
+	name = "QuikDeploy Barricade"
+	desc = "This is a QuikDeploy barricade, allows for extremely fast laying of barricades. Usually found in specific vendors and cannot be picked back up."
+	icon = 'icons/obj/items/quikdeploy_cade.dmi'
+	w_class = WEIGHT_CLASS_SMALL //While this is small, normal 50 stacks of metal is NORMAL so this is a bit on the bad space to cade ratio
+	var/delay = 0 //Delay on deploying the thing
+	var/atom/thing_to_deploy = /obj
```
Comment (Rohesie) [inline code/game/objects/items/quickdeploy_cade.dm]:
```suggestion
	var/atom/thing_to_deploy = null
```
Base of `obj` will error silently.

File: code/game/objects/items/quickdeploy_cade.dm [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+/obj/item/quikdeploy/examine(mob/user)
+	. = ..()
+	to_chat(user, "This QuikDeploy cade seems to deploy a [thing_to_deploy.name].")
```
Comment (Rohesie) [inline code/game/objects/items/quickdeploy_cade.dm]:
```suggestion
	to_chat(user, "This QuikDeploy system seems to deploy \a [thing_to_deploy.name].")
```

File: code/game/objects/items/quickdeploy_cade.dm [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+		qdel(src)
+
+/obj/item/quikdeploy/proc/can_place(mob/user)
+	to_chat(user, "This thing doesn't actually really do anything! complain to whoever gave you this")
+	return FALSE
+
+/obj/item/quikdeploy/cade
+	thing_to_deploy = /obj/structure/barricade/metal
+	icon_state = "metal"
+	delay = 30 //3 seconds default
```
Comment (Rohesie) [inline code/game/objects/items/quickdeploy_cade.dm]:
```suggestion
	delay = 3 SECONDS
```

File: code/game/objects/items/quickdeploy_cade.dm [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+		to_chat(user, "You decide against deploying something here.")
+		return
+	if(can_place(user)) //can_place() handles sending the error and success messages to the user
+		var/obj/O = new thing_to_deploy(get_turf(user))
+		O.setDir(user.dir)
+		playsound(loc, 'sound/items/ratchet.ogg', 25, TRUE)
+		qdel(src)
+
+/obj/item/quikdeploy/proc/can_place(mob/user)
+	to_chat(user, "This thing doesn't actually really do anything! complain to whoever gave you this")
```
Comment (Rohesie) [inline code/game/objects/items/quickdeploy_cade.dm]:
```suggestion
	to_chat(user, "This thing doesn't seem to actually do anything. Complain to whoever gave you this!")
```

Comment (Rohesie) [reply code/game/objects/items/quickdeploy_cade.dm]: Add proper classes to text feedback.

File: code/game/objects/items/quickdeploy_cade.dm [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+		if(!thing.density) //not dense, move on
+			continue
+		if(!(thing.flags_atom & ON_BORDER)) //dense and non-directional, end
+			to_chat(user, "<span class='warning'>No space here for a barricade.</span>")
+			return FALSE
+		if(thing.dir != user.dir)
+			continue
+		to_chat(user, "<span class='warning'>No space here for a barricade.</span>")
+		return FALSE
+	to_chat(user, "You plop down the barricade in front of you.")
```
Comment (Rohesie) [inline code/game/objects/items/quickdeploy_cade.dm]:
```suggestion
	to_chat(user, "<span class='notice'>You plop down the barricade in front of you.</span>")
```

File: code/game/objects/items/quickdeploy_cade.dm [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+	var/atom/thing_to_deploy = /obj
+
+/obj/item/quikdeploy/examine(mob/user)
+	. = ..()
+	to_chat(user, "This QuikDeploy cade seems to deploy a [thing_to_deploy.name].")
+
+/obj/item/quikdeploy/attack_self(mob/user)
+	to_chat(user, "You start to deploy onto the tile in front of you...")
+	if(!do_after(usr, delay, TRUE, src, BUSY_ICON_BUILD))
+		to_chat(user, "You decide against deploying something here.")
```
Comment (Rohesie) [inline code/game/objects/items/quickdeploy_cade.dm]: `<span class='warning'>`

File: code/game/objects/items/quickdeploy_cade.dm [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+	w_class = WEIGHT_CLASS_SMALL //While this is small, normal 50 stacks of metal is NORMAL so this is a bit on the bad space to cade ratio
+	var/delay = 0 //Delay on deploying the thing
+	var/atom/thing_to_deploy = /obj
+
+/obj/item/quikdeploy/examine(mob/user)
+	. = ..()
+	to_chat(user, "This QuikDeploy cade seems to deploy a [thing_to_deploy.name].")
+
+/obj/item/quikdeploy/attack_self(mob/user)
+	to_chat(user, "You start to deploy onto the tile in front of you...")
```
Comment (Rohesie) [inline code/game/objects/items/quickdeploy_cade.dm]: <span class='notice'>

File: code/game/objects/items/quickdeploy_cade.dm [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+	if(can_place(user)) //can_place() handles sending the error and success messages to the user
+		var/obj/O = new thing_to_deploy(get_turf(user))
+		O.setDir(user.dir)
+		playsound(loc, 'sound/items/ratchet.ogg', 25, TRUE)
+		qdel(src)
+
+/obj/item/quikdeploy/proc/can_place(mob/user)
+	if(isnull(thing_to_deploy)) //Spaghetti or wrong type spawned
+		to_chat(user, "<span class='warning'>This thing doesn't actually really do anything! Complain to whoever gave you this")
+		return FALSE
```
Comment (Rohesie) [inline code/game/objects/items/quickdeploy_cade.dm]:
```suggestion
```

File: code/game/objects/items/quickdeploy_cade.dm [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+/obj/item/quikdeploy/cade
+	thing_to_deploy = /obj/structure/barricade/metal
+	icon_state = "metal"
+	delay = 3 SECONDS
+
+/obj/item/quikdeploy/cade/can_place(mob/user)
+	. = ..()
+	if(!.)
+		return FALSE
+	for(var/obj/thing in user.loc)
```
Comment (Rohesie) [inline code/game/objects/items/quickdeploy_cade.dm]: Add something like this here:
```
if(!isturf(user.loc))
	return FALSE
var/turf/placement_loc = user.loc
if(!isturf(placement_loc.density))
	return FALSE

**Pull request merged**